src/ImageSharp/Formats/Tiff/Writers/TiffStreamWriter.cs

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Six Labors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.IO;
+
+namespace SixLabors.ImageSharp.Formats.Experimental.Tiff.Writers
+{
+    /// <summary>
+    /// Utility class for writing TIFF data to a <see cref="Stream"/>.
+    /// </summary>
+    internal class TiffStreamWriter : IDisposable
+    {
+        private static readonly byte[] PaddingBytes = new byte[4];
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TiffStreamWriter"/> class.
+        /// </summary>
+        /// <param name="output">The output stream.</param>
+        public TiffStreamWriter(Stream output) => this.BaseStream = output;
+
+        /// <summary>
+        /// Gets a value indicating whether the architecture is little-endian.
+        /// </summary>
+        public bool IsLittleEndian => BitConverter.IsLittleEndian;
+
+        /// <summary>
+        /// Gets the current position within the stream.
+        /// </summary>
+        public long Position => this.BaseStream.Position;
+
+        /// <summary>
+        /// Gets the base stream.
+        /// </summary>
+        public Stream BaseStream { get; }
+
+        /// <summary>
+        /// Writes an empty four bytes to the stream, returning the offset to be written later.
+        /// </summary>
+        /// <returns>The offset to be written later</returns>
+        public long PlaceMarker()
+        {
+            long offset = this.BaseStream.Position;
+            this.Write(0u);
+            return offset;
+        }
+
+        /// <summary>
+        /// Writes an array of bytes to the current stream.
+        /// </summary>
+        /// <param name="value">The bytes to write.</param>
+        public void Write(byte[] value) => this.BaseStream.Write(value, 0, value.Length);
+
+        /// <summary>
+        /// Writes the specified value.
+        /// </summary>
+        /// <param name="value">The bytes to write.</param>
+        public void Write(ReadOnlySpan<byte> value) => this.BaseStream.Write(value);
+
+        /// <summary>
+        /// Writes a byte to the current stream.
+        /// </summary>
+        /// <param name="value">The byte to write.</param>
+        public void Write(byte value) => this.BaseStream.Write(new byte[] { value }, 0, 1);
+
+        /// <summary>
+        /// Writes a two-byte unsigned integer to the current stream.
+        /// </summary>
+        /// <param name="value">The two-byte unsigned integer to write.</param>
+        public void Write(ushort value)
+        {
+            byte[] bytes = BitConverter.GetBytes(value);
```

Comment:
`BitConverter` is heavy on allocations. We should create a version that defers to `BinaryPrimitives` and allows using `Span<byte>`

Reply:
I have changed the TiffStreamWriter to use BinaryPrimitives and a scratch buffer to reduce allocations with: b5882534